packages/lesswrong/server/fmCrosspost.ts

```diff
@@ -0,0 +1,381 @@
+  }
+}
+
+const getSecret = () => {
+  const secret = crosspostSigningKeySetting.get();
+  if (!secret) {
+    throw new MissingSecretError();
+  }
+  return secret;
+}
+
+const getUserId = (req?: Request) => {
+  const userId = req?.user?._id;
+  if (!userId) {
+    throw new UnauthorizedError();
+  }
+  return userId;
+}
+
+const makeApiUrl = (route: ApiRoute) => fmCrosspostBaseUrlSetting.get() + route.slice(1);
+
+const makeCrossSiteRequest = async <T extends {}>(
+  route: ApiRoute,
+  body: T,
+  expectedStatus: string,
+  onErrorMessage: string,
+) => {
+  const result = await fetch(makeApiUrl(route), {
+    method: "POST",
+    headers: {
+      "Content-Type": "application/json",
+    },
+    body: JSON.stringify(body),
+  });
+  const json = await result.json();
+  if (json.status !== expectedStatus) {
+    throw new ApiError(500, onErrorMessage);
+  }
+  return json;
+}
+
+const signToken = <T extends {}>(payload: T): string =>
+  jwt.sign(payload, getSecret(), jwtSigningOptions);
+
+const verifyToken = <T extends {}>(token: string): T =>
+  jwt.verify(token, getSecret()) as T;
+
+const crosspostResolvers = {
+  Mutation: {
+    connectCrossposter: async (
+      root: void,
+      {token}: ConnectCrossposterArgs,
+      {req, res}: ResolverContext,
+    ) => {
+      const localUserId = getUserId(req);
+      const {foreignUserId} = await makeCrossSiteRequest(
+        apiRoutes.connectCrossposter,
+        {token, localUserId},
+        "connected",
+        "Failed to connect accounts for crossposting",
+      );
+      await Users.rawUpdateOne({_id: localUserId}, {
+        $set: {fmCrosspostUserId: foreignUserId},
+      });
+      return "success";
+    },
+    unlinkCrossposter: async (root: void, args: {}, {req, res}: ResolverContext) => {
+      const localUserId = getUserId(req);
+      const foreignUserId = req?.user?.fmCrosspostUserId;
+      if (foreignUserId) {
+        const token = signToken<UnlinkCrossposterPayload>({userId: foreignUserId});
+        await makeCrossSiteRequest(
+          apiRoutes.unlinkCrossposter,
+          {token},
+          "unlinked",
+          "Failed to unlink crossposting accounts",
+        );
+        await Users.rawUpdateOne({_id: localUserId}, {
+          $unset: {fmCrosspostUserId: ""},
+        });
+      }
+      return "success";
+    },
+  },
+};
+
+addGraphQLResolvers(crosspostResolvers);
+addGraphQLMutation("connectCrossposter(token: String): String");
+addGraphQLMutation("unlinkCrossposter: String");
+
+const updateCrosspost = async (post: DbPost) => {
+  if (post.fmCrosspost?.foreignPostId) {
+    const token = signToken<UpdateCrosspostPayload>({
+      postId: post.fmCrosspost.foreignPostId,
+      draft: post.draft,
+      deletedDraft: post.deletedDraft,
+      title: post.title,
+    });
+    await makeCrossSiteRequest(
+      apiRoutes.updateCrosspost,
+      {token},
+      "updated",
+      "Failed to update crosspost draft status",
+    );
+  }
```

Comment:
If the entire body of the function is in an `if` block, and it's otherwise a no-op, I'm tempted to surface that check to the caller, especially since we're already checking a few other things before calling it.